### PR TITLE
to get sherpa to run Apple M1 (arm)

### PR DESCRIPTION
**Pull request by dtnguyen2**
# Note

This PR enables sherpa to run on Apple M1, without XSpec.

```
=============================== warnings summary ===============================
../../../../proj/port_ots/macOS-BigSur-ARM/Python-3.9.4.macOS-BigSur-ARM/lib/python3.9/site-packages/_pytest/cacheprovider.py:428
  /proj/port_ots/macOS-BigSur-ARM/Python-3.9.4.macOS-BigSur-ARM/lib/python3.9/site-packages/_pytest/cacheprovider.py:428: PytestCacheWarning: could not create cache path /.pytest_cache/v/cache/nodeids
    config.cache.set("cache/nodeids", sorted(self.cached_nodeids))

../../../../proj/port_ots/macOS-BigSur-ARM/Python-3.9.4.macOS-BigSur-ARM/lib/python3.9/site-packages/_pytest/stepwise.py:49
  /proj/port_ots/macOS-BigSur-ARM/Python-3.9.4.macOS-BigSur-ARM/lib/python3.9/site-packages/_pytest/stepwise.py:49: PytestCacheWarning: could not create cache path /.pytest_cache/v/cache/stepwise
    session.config.cache.set(STEPWISE_CACHE_DIR, [])

-- Docs: https://docs.pytest.org/en/stable/warnings.html
===== 3705 passed, 924 skipped, 71 xfailed, 2 warnings in 99.23s (0:01:39) =====
```

### CODE REVIEW

File: sherpa/utils/__init__.py
Comment (DougBurke) [inline sherpa/utils/__init__.py:1325]: I think it would be better to just check whether the attributes exist rather than base it on the platform type (e.g. windows may not have these symbols, and whilst we currently don't support a windows build we should try to avoid hard-coding logic when we can):

```python
try:
    converters[numpy.complex256] = 'Complex256'
except AttributeError:
    pass

try:
    converters[numpy.float128] = 'Float128'
except AttributeError:
    pass
```

Note above code is untested...

**Comment by DougBurke** [comment]
Are we waiting for SDS review on this?

**Comment by dtnguyen2** [comment]
Still working on this

File: sherpa/utils/__init__.py
Comment (dtnguyen2) [reply sherpa/utils/__init__.py]: made the changes you requested

Comment (DougBurke) [reply sherpa/utils/__init__.py]: Ta

**Review by DougBurke — APPROVED** [review]
FWIW LGTM (I am just assuming that the defines make sense as we currently don't have any CI to test this out).

File: sherpa/utils/src/cephes/const.c
Comment (wmclaugh) [inline sherpa/utils/src/cephes/const.c:92]: The preprecessor conditionals here change the behavior compared to the original code due to the nesting. Is this intended?

in the original-  if INFINITIES is not defined, INFINITY is set to 1.79...

    #ifdef INFINITIES
        double INFINITY = strtod("INF");  /* 99e999; */
    #else
        double INFINITY =  1.79769313486231570815E308;    /* 2**1024*(1-MACHEP) */
    #endif

in the modified code- if INFINITIES is not defined, nothing happens...

    #ifdef INFINITIES
        #ifndef INFINITY
            double INFINITY = 1.0/0.0;
        #else
            double INFINITY =  1.79769313486231570815E308;    /* 2**1024*(1-MACHEP) */
        #endif
    #endif

Comment (dtnguyen2) [reply sherpa/utils/src/cephes/const.c:92]: The following simple program will test if one can use ~1.79e308 and 0 for infinity and NaN, respectively; Moreover a check if 1.0/0.0 and 1.0/0. - 1/0.0 are legit definitions for infinity and NaN:

```
devel12-107: cat tst_inf_nan.c 
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

/*double tmp = sqrt(-1); */


double myinf = (1.0/0.0);
double not_inf = 1.79769313486231570815E308;
double mynan = (1.0/0. * 0.0);
double not_nan = 0.0;

int main() {

  printf("myinf = %g\tis myinf inf: %d\n", myinf, isinf(myinf));
  printf("not_inf = %g\tis not_inf inf: %d\n", not_inf, isinf(not_inf));

  printf("mynan = %g\tis mynan nan: %d\n", mynan, isnan(mynan));
  printf("not_nan = %g\tis not_nan nan: %d\n", not_nan, isnan(not_nan));

  return 0;
}
```

As you can see, if one uses the builtin ```isinf``` and ```isnan`` to test out a few numbers:

```
devel12-108: gcc tst_inf_nan.c 
devel12-109: a.out
myinf = inf	is myinf inf: 1
not_inf = 1.79769e+308	is not_inf inf: 0
mynan = nan	is mynan nan: 1
not_nan = 0	is not_nan nan: 0
```
QED


There are two problems with the line:
```
double INFINITY = strtod("INF");  /* 99e999; */
```
1) The function ```strtod``` takes two arguments:
```
NAME
       strtod, strtof, strtold - convert ASCII string to floating-point number

SYNOPSIS
       #include <stdlib.h>

       double strtod(const char *nptr, char **endptr);
       float strtof(const char *nptr, char **endptr);
       long double strtold(const char *nptr, char **endptr);
```
2) modern compilers will only allow one to call a function from within a function.  If I comment out the call to the ```sqrt``` function like:

```
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

double tmp = sqrt(-1);


double myinf = (1.0/0.0);
double not_inf = 1.79769313486231570815E308;
double mynan = (1.0/0. * 0.0);
double not_nan = 0.0;

int main() {

  printf("myinf = %g\tis myinf inf: %d\n", myinf, isinf(myinf));
  printf("not_inf = %g\tis not_inf inf: %d\n", not_inf, isinf(not_inf));

  printf("mynan = %g\tis mynan nan: %d\n", mynan, isnan(mynan));
  printf("not_nan = %g\tis not_nan nan: %d\n", not_nan, isnan(not_nan));

  return 0;
}
```
So I can't compile it:

```
devel12-115: gcc tst_inf_nan.c
tst_inf_nan.c:5:1: error: initializer element is not constant
 double tmp = sqrt(-1);
 ^
devel12-116: 
```

**Review by harlantc — APPROVED** [review]
This looks good to me.

**Comment by dtnguyen2** [comment]
Just realized that when I made the changes: checking if the platform is an ```arm``` as suggested by @DougBurke then I probably should have deleted the following import:
```
import platform
```

**Pull request merged**
